gq/gq_iso_test.go

```diff
@@ -0,0 +1,162 @@
+const sigHex = "99394F1D15924C0374CF80C7274CD9F232903A6423D9327156F69743EAEF03E1EFEDFDA8474C97F6570D9EF53C6CE2AE2BA68D01FFF9AA82068214BCD775B95CC297DDC38A63741AB3166B58275E0FB728D26DB18A2C3F14B621CF3863F8648B3149FE896348BE73D37E2F06E6E26C84C044984C09C658300B58EC2383E3B0A1F1390D62B772A69F37B5"
+
+var mISO = []byte("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopqo")
+var sigISO []byte
+var qISO []byte
+var idISO []byte
+var svISO signerVerifier
+
+// Test our signer using the values specified in ISO/IEC 14888-2:2008
+func TestSignerISO(t *testing.T) {
+	initializeISOValues(t)
+
+	// The test vector specifies that h(W||M) use SHA-1
+	h := hash
+	hash = sha1Hash
+
+	// Instead of a true random value for r, use the value from the test vector
+	rn := randomNumbers
+	randomNumbers = hardcodedRandomISO
+
+	// restore default function definitions after the test
+	defer func() {
+		hash = h
+		randomNumbers = rn
+	}()
+
+	encodedSig, err := svISO.Sign(qISO, mISO)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sig, err := util.Base64DecodeForJWT(encodedSig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !bytes.Equal(sig, sigISO) {
+		t.Fatal("Signature does not match expected value")
+	}
+}
+
+// Test our verifier using the values specified in ISO/IEC 14888-2:2008
+func TestVerifierISO(t *testing.T) {
+	initializeISOValues(t)
+
+	// The test vector specifies that h(W||M) use SHA-1
+	h := hash
+	hash = sha1Hash
+
+	// The test vector formats Id with PSS. Instead of implementing this
+	// encoding ourselves, we hardcode the value for G given in the standard
+	ep := encodePKCS1v15
+	encodePKCS1v15 = pssEncodedId
+
+	// restore default function definitions after the test
+	defer func() {
+		hash = h
+		encodePKCS1v15 = ep
+	}()
+
+	encodedSigISO := util.Base64EncodeForJWT(sigISO)
+	ok := svISO.Verify(encodedSigISO, idISO, mISO)
+
+	if !ok {
+		t.Fatal("signature verification failed")
+	}
+}
+
+func initializeISOValues(t *testing.T) {
+	vBytes, err := hexToBytes(vHex)
+	if err != nil {
+		t.Fatal(err)
+	}
+	v := new(big.Int).SetBytes(vBytes)
+
+	nBytes, err := hexToBytes(nHex)
+	if err != nil {
+		t.Fatal(err)
+	}
+	n, err := bigmod.NewModulusFromBig(new(big.Int).SetBytes(nBytes))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	svISO = signerVerifier{n, v, 128, 10, 1}
+
+	sigISO, err = hexToBytes(sigHex)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	qISO, err = hexToBytes(qHex)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	idISO, err = hexToBytes(idHex)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+var pssEncodedId = func(k int, data []byte) []byte {
+	em, _ := hexToBytes(gHex)
+	return em
+}
+
+var hardcodedRandomISO = func(t int, n *bigmod.Modulus) ([]*bigmod.Nat, error) {
+	ys := make([]*bigmod.Nat, t)
+
+	rRaw, err := hexToBytes(rHex)
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := bigmod.NewNat().SetBytes(rRaw, n)
+	if err != nil {
+		return nil, err
+	}
+
+	ys[0] = r
+	return ys, nil
+}
```

Comment:
Not sure I fully understand this part. `ys` is initialized to have length `t` but we only fill out the first one, so the rest are nil. Should the length just be 1?

Reply:
Good callout, thanks!